File: clients/js/schema-client.js
/**
 * SchemaClient — JS helper for running the proxy's saved requests and checking
 * what came back.
 *
 * The other three clients in this folder stage a condition; this one **asserts
 * an outcome**. A saved request can carry an expectation — an expected status,
 * a JSON Schema for the body, or both — and this runs it through the proxy and
 * tells you whether the response met it:
 *
 *   const { SchemaClient } = require("./schema-client");
 *   const client = new SchemaClient();
 *
 *   // Throws, listing every problem, if the response didn't match:
 *   await client.assertPasses("get-order");
 *
 *   // A whole collection, in order, one at a time:
 *   await client.assertCollectionPasses("checkout-flow");
 *
 *   // Or with a schema kept in *your* repo, next to this test:
 *   await client.assertPasses("get-order", {
 *     expect: { status: 200, schema: require("./schemas/order.json") },
 *   });
 *
 * **The schema is never evaluated here.** Every check is answered by the same
 * `utils/schema-validate.js` the dashboard uses, so this file cannot drift from
 * what the proxy actually enforces — the same reason `BlockClient` asks the
 * server whether a path is blocked instead of re-deriving the prefix rule.
 *
 * **A request with no expectation does not pass.** `assertPasses` throws on
 * one, because an assertion that checked nothing and returned green is the
 * failure this whole feature exists to prevent. Pass your own `expect`, or
 * construct with `{ requireCheck: false }` if you genuinely mean "just send
 * it".
 *
 * Standalone on purpose: zero dependencies (node:http only) so this file can be
 * copied into any QA repo. The port autodetection below intentionally
 * duplicates mock-client.js / capture-client.js / block-client.js /
 * scripts/cli.js — keeping the client droppable beats sharing.
 */
const http = require("http");

const PREFERRED_PORT = 8888; // proxy's preferred port (often already taken)
const PORT_SCAN = 20; // matches the proxy's own fallback scan range

/** Probe a port for OUR admin server (another proxy on 8888 won't match). */
function _probe(host, port) {
  return new Promise((resolve) => {
    const req = http.request(
      { hostname: host, port, path: "/__admin/health", method: "GET", timeout: 500 },
      (res) => {
        let data = "";
        res.on("data", (ch) => (data += ch));
        res.on("end", () => {
          try {
            const j = JSON.parse(data);
            resolve(!!j && j.status === "ok" && Array.isArray(j.instances));
          } catch {
            resolve(false);
          }
        });
      }
    );
    req.on("error", () => resolve(false));
    req.on("timeout", () => {
      req.destroy();
      resolve(false);
    });
    req.end();
  });
}

/**
 * A failed expectation. Carries the errors as a list so a test runner can show
 * them one per line instead of one long string.
 */
class ExpectationFailed extends Error {
  constructor(label, errors) {
    super(`${label} failed its check:\n  ${errors.join("\n  ")}`);
    this.name = "ExpectationFailed";
    this.errors = errors;
  }
}

class SchemaClient {
  /**
   * @param {{ host?: string, port?: number, requireCheck?: boolean }} [opts] —
   *   explicit port skips autodetection (recommended in hermetic tests).
   *   Defaults honor the MOCK_HOST / MOCK_PORT environment variables.
   *   `requireCheck: false` lets `assertPasses` accept a request that checks
   *   nothing, instead of throwing to say the assertion was hollow.
   */
  constructor({ host, port, requireCheck = true } = {}) {
    this.host = host || process.env.MOCK_HOST || "localhost";
    this.port =
      port || (process.env.MOCK_PORT ? parseInt(process.env.MOCK_PORT, 10) : null);
    this.requireCheck = requireCheck !== false;
  }

  async _resolvePort() {
    if (this.port) return this.port;
    for (let p = PREFERRED_PORT; p <= PREFERRED_PORT + PORT_SCAN; p++) {
      if (await _probe(this.host, p)) return (this.port = p);
    }
    throw new Error(
      `SchemaClient: no mock proxy found on ${this.host}:${PREFERRED_PORT}-${
        PREFERRED_PORT + PORT_SCAN
      } (is the server running? set MOCK_PORT to override)`
    );
  }

  async _request(method, path, body) {
    const port = await this._resolvePort();
    const payload = body === undefined ? null : JSON.stringify(body);
    const { status, data } = await new Promise((resolve, reject) => {
      const headers = { "Content-Type": "application/json" };
      if (payload) headers["Content-Length"] = Buffer.byteLength(payload);
      const req = http.request(
        { hostname: this.host, port, path: `/__admin${path}`, method, headers },
        (res) => {
          let raw = "";
          res.on("data", (ch) => (raw += ch));
          res.on("end", () => resolve({ status: res.statusCode, data: raw }));
        }
      );
      req.on("error", reject);
      if (payload) req.write(payload);
      req.end();
    });
    let parsed;
    try {
      parsed = JSON.parse(data);
    } catch {
      parsed = data;
    }
    if (status < 200 || status >= 300) {
      const msg = (parsed && parsed.error) || data || `HTTP ${status}`;
      throw new Error(`SchemaClient: ${method} ${path} failed (${status}): ${msg}`);
    }
    return parsed;
  }

  /** Every saved request, newest first — id, name, path, and its `expect`. */
  async savedRequests() {
    return (await this._request("GET", "/saved-requests")).requests || [];
  }

  /** Every collection, ids already resolved into whole records. */
  async collections() {
    return (await this._request("GET", "/collections")).collections || [];
  }

  /**
   * Which saved requests actually check their response.
   *
   * Worth asking out loud before trusting a green run: a suite of requests that
   * assert nothing passes every time.
   *
   * @returns {Promise<Array<{id, name, path, checked: boolean, expect}>>}
   */
  async checks() {
    return (await this.savedRequests()).map((r) => ({
      id: r.id,
      name: r.name,
      method: r.method || "GET",
      path: r.path,
      checked: Boolean(r.expect),
      expect: r.expect || null,
    }));
  }

  /**
   * Send one saved request through the proxy and report what came back.
   *
   * Does **not** throw on a failed expectation — that is `assertPasses`. This
   * is for when you want the verdict as data.
   *
   * @param {string} id — a saved request id (see `checks()`)
   * @param {{expect?: object|null}} [opts] — `expect` **replaces** the stored
   *   expectation for this call, which is how you keep schemas in your own repo
   *   next to the tests that use them. `expect: null` sends without checking.
   * @returns {Promise<{status: number, checked: boolean, passed: boolean,
   *   errors: string[]}>}
   */
  async run(id, opts = {}) {
    const body = Object.hasOwn(opts, "expect") ? { expect: opts.expect } : {};
    const res = await this._request(
      "POST",
      `/saved-requests/${encodeURIComponent(id)}/send`,
      body
    );
    return {
      status: res.status,
      checked: Boolean(res.expect),
      // `passed` is only meaningful when something was checked; `checked` is
      // what tells the two apart, and callers must not conflate them.
      passed: res.expect ? res.expect.passed : false,
      errors: res.expect ? res.expect.errors : [],
    };
  }

  /**
   * Send one saved request and throw unless the response met its expectation.
   *
   * Throws `ExpectationFailed` (with `.errors`) when the check failed, and a
   * plain Error when the request **had no check at all** — see `requireCheck`.
   *
   * @returns {Promise<object>} the same result `run()` gives, on success
   */
  async assertPasses(id, opts = {}) {
    const result = await this.run(id, opts);
    if (!result.checked) {
      if (!this.requireCheck) return result;
      throw new Error(
        `SchemaClient: saved request "${id}" has no expectation, so this assertion ` +
          `checked nothing and would pass whatever came back. Add one from the ` +
          `dashboard's Expect tab, pass your own { expect: … }, or construct ` +
          `with { requireCheck: false } if sending without checking is intended.`
      );
    }
    if (!result.passed) throw new ExpectationFailed(`"${id}"`, result.errors);
    return result;
  }

  /**
   * Run a whole collection, **in order, one at a time**.
   *
   * Sequential and awaited because that is the shape these have — "log in, then
   * call the thing that needs the token". It does not stop at the first
   * failure: a run is how you find out *where* a flow breaks, and the results
   * after the red one are part of that answer.
   *
   * @param {string} nameOrId
   * @returns {Promise<Array<{id, name, status, checked, passed, errors}>>}
   */
  async runCollection(nameOrId) {
    const groups = await this.collections();
    const group = groups.find((g) => g.id === nameOrId || g.name === nameOrId);
    if (!group) {
      const known = groups.map((g) => g.id).join(", ") || "none";
      throw new Error(`SchemaClient: no collection "${nameOrId}". Known: ${known}`);
    }

    const results = [];
    for (const record of group.requests) {
      let outcome;
      try {
        outcome = await this.run(record.id);
      } catch (err) {
        // A request that could not be sent at all — SSL off for its target, its
        // instance gone, a stored schema that cannot be honoured. It belongs in
        // the results as a failure, not as a thrown exception that hides the
        // rows after it.
        outcome = { status: 0, checked: true, passed: false, errors: [err.message] };
      }
      results.push({ id: record.id, name: record.name, ...outcome });
    }
    return results;
  }

  /**
   * Run a collection and throw unless **every** request met its expectation.
   *
   * The message names each failure with its request, because "something in
   * checkout-flow broke" is not an answer anybody can act on.
   *
   * Unchecked requests are reported the same way `assertPasses` treats one:
   * they are not passes. Turn that off with `{ requireCheck: false }`.
   */
  async assertCollectionPasses(nameOrId) {
    const results = await this.runCollection(nameOrId);

    const problems = [];
    results.forEach((r) => {
      if (!r.checked) {
        if (this.requireCheck) problems.push(`${r.name}: nothing was checked`);
        return;
      }
      if (!r.passed) r.errors.forEach((line) => problems.push(`${r.name}: ${line}`));
    });

    if (problems.length)
      throw new ExpectationFailed(`Collection "${nameOrId}"`, problems);
    return results;
  }
}

module.exports = { SchemaClient, ExpectationFailed };

File: clients/python/schema_client.py
"""SchemaClient -- Python helper for running the proxy's saved requests and
checking what came back.

The other three clients in this folder stage a condition; this one **asserts an
outcome**. A saved request can carry an expectation -- an expected status, a
JSON Schema for the body, or both -- and this runs it through the proxy and
tells you whether the response met it:

    from schema_client import SchemaClient, ExpectationFailed

    client = SchemaClient()

    # Raises, listing every problem, if the response didn't match:
    client.assert_passes("get-order")

    # A whole collection, in order, one at a time:
    client.assert_collection_passes("checkout-flow")

    # Or with a schema kept in *your* repo, next to this test:
    client.assert_passes("get-order", expect={
        "status": 200,
        "schema": json.load(open("schemas/order.json")),
    })

**The schema is never evaluated here.** Every check is answered by the same
utils/schema-validate.js the dashboard uses, so this file cannot drift from what
the proxy actually enforces -- the same reason BlockClient asks the server
whether a path is blocked instead of re-deriving the prefix rule.

**A request with no expectation does not pass.** assert_passes() raises on one,
because an assertion that checked nothing and returned green is the failure this
whole feature exists to prevent. Pass your own `expect`, or construct with
require_check=False if you genuinely mean "just send it".

Standalone on purpose: stdlib only, so this file can be copied into any QA repo.
The port autodetection below intentionally duplicates mock_client.py /
capture_client.py / block_client.py / scripts/cli.js -- keeping the client
droppable beats sharing code.
"""

import json
import os
import urllib.error
import urllib.parse
import urllib.request

PREFERRED_PORT = 8888  # proxy's preferred port (often already taken)
PORT_SCAN = 20  # matches the proxy's own fallback scan range

# Distinguishes "no expect argument" from "expect=None", which means the
# opposite thing: send this request and check nothing at all.
_UNSET = object()


class SchemaError(Exception):
    """Raised when the run API returns an error response."""


class ExpectationFailed(SchemaError):
    """Raised when a response did not match what was expected.

    Its own type because it is the one failure worth catching separately: the
    request went out and came back fine, the *answer* was wrong. `.errors` is
    the list, so a test runner can print one per line.
    """

    def __init__(self, label, errors):
        super().__init__("{} failed its check:\n  {}".format(label, "\n  ".join(errors)))
        self.errors = list(errors)


class NothingCheckedError(SchemaError):
    """Raised when an assertion was made about a request that checks nothing.

    Separate from a failure because it is the opposite problem: nothing went
    wrong, and that is exactly why you cannot trust the green.
    """


def _probe(host, port):
    """Probe a port for OUR admin server (another proxy on 8888 won't match)."""
    url = "http://{}:{}/__admin/health".format(host, port)
    try:
        with urllib.request.urlopen(url, timeout=0.5) as res:
            body = json.loads(res.read().decode("utf-8"))
            return body.get("status") == "ok" and isinstance(body.get("instances"), list)
    except Exception:
        return False


class SchemaClient:
    """Run saved requests through the proxy and assert on the responses.

    An explicit `port` skips autodetection (recommended in hermetic tests).
    Defaults honor the MOCK_HOST / MOCK_PORT environment variables.
    `require_check=False` lets assert_passes() accept a request that checks
    nothing, instead of raising to say the assertion was hollow.
    """

    def __init__(self, host=None, port=None, require_check=True):
        self.host = host or os.environ.get("MOCK_HOST", "localhost")
        env_port = os.environ.get("MOCK_PORT")
        self.port = port or (int(env_port) if env_port else None)
        self.require_check = require_check

    def _resolve_port(self):
        if self.port:
            return self.port
        for candidate in range(PREFERRED_PORT, PREFERRED_PORT + PORT_SCAN + 1):
            if _probe(self.host, candidate):
                self.port = candidate
                return candidate
        raise SchemaError(
            "SchemaClient: no mock proxy found on {}:{}-{} "
            "(is the server running? set MOCK_PORT to override)".format(
                self.host, PREFERRED_PORT, PREFERRED_PORT + PORT_SCAN
            )
        )

    def _request(self, method, path, body=_UNSET):
        port = self._resolve_port()
        url = "http://{}:{}/__admin{}".format(self.host, port, path)
        data = None if body is _UNSET else json.dumps(body).encode("utf-8")
        req = urllib.request.Request(
            url, data=data, method=method, headers={"Content-Type": "application/json"}
        )
        try:
            with urllib.request.urlopen(req, timeout=35) as res:
                return json.loads(res.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raw = exc.read().decode("utf-8", "replace")
            try:
                message = json.loads(raw).get("error", raw)
            except ValueError:
                message = raw
            raise SchemaError(
                "SchemaClient: {} {} failed ({}): {}".format(
                    method, path, exc.code, message
                )
            ) from exc
        except urllib.error.URLError as exc:
            raise SchemaError(
                "SchemaClient: could not reach {}: {}".format(url, exc.reason)
            ) from exc

    def saved_requests(self):
        """Every saved request, newest first -- id, name, path, and its expect."""
        return self._request("GET", "/saved-requests").get("requests", [])

    def collections(self):
        """Every collection, ids already resolved into whole records."""
        return self._request("GET", "/collections").get("collections", [])

    def checks(self):
        """Which saved requests actually check their response.

        Worth asking out loud before trusting a green run: a suite of requests
        that assert nothing passes every time.
        """
        return [
            {
                "id": r.get("id"),
                "name": r.get("name"),
                "method": r.get("method") or "GET",
                "path": r.get("path"),
                "checked": bool(r.get("expect")),
                "expect": r.get("expect"),
            }
            for r in self.saved_requests()
        ]

    def run(self, request_id, expect=_UNSET):
        """Send one saved request through the proxy and report what came back.

        Does **not** raise on a failed expectation -- that is assert_passes().
        This is for when you want the verdict as data.

        `expect` **replaces** the stored expectation for this call, which is how
        you keep schemas in your own repo next to the tests that use them.
        `expect=None` sends without checking anything.

        Returns {"status", "checked", "passed", "errors"}.
        """
        body = {} if expect is _UNSET else {"expect": expect}
        res = self._request(
            "POST",
            "/saved-requests/{}/send".format(urllib.parse.quote(str(request_id), safe="")),
            body,
        )
        verdict = res.get("expect")
        return {
            "status": res.get("status"),
            "checked": bool(verdict),
            # "passed" is only meaningful when something was checked; "checked"
            # is what tells the two apart, and callers must not conflate them.
            "passed": bool(verdict and verdict.get("passed")),
            "errors": list(verdict.get("errors", [])) if verdict else [],
        }

    def assert_passes(self, request_id, expect=_UNSET):
        """Send one saved request and raise unless the response met its check.

        Raises ExpectationFailed (with .errors) when the check failed, and
        NothingCheckedError when the request had no check at all.
        """
        result = self.run(request_id, expect)
        if not result["checked"]:
            if not self.require_check:
                return result
            raise NothingCheckedError(
                'SchemaClient: saved request "{}" has no expectation, so this '
                "assertion checked nothing and would pass whatever came back. Add "
                "one from the dashboard's Expect tab, pass your own expect=..., or "
                "construct with require_check=False if sending without checking is "
                "intended.".format(request_id)
            )
        if not result["passed"]:
            raise ExpectationFailed('"{}"'.format(request_id), result["errors"])
        return result

    def run_collection(self, name_or_id):
        """Run a whole collection, **in order, one at a time**.

        Sequential because that is the shape these have -- "log in, then call
        the thing that needs the token". It does not stop at the first failure:
        a run is how you find out *where* a flow breaks, and the results after
        the red one are part of that answer.
        """
        groups = self.collections()
        group = next(
            (g for g in groups if g.get("id") == name_or_id or g.get("name") == name_or_id),
            None,
        )
        if group is None:
            known = ", ".join(g.get("id", "?") for g in groups) or "none"
            raise SchemaError(
                'SchemaClient: no collection "{}". Known: {}'.format(name_or_id, known)
            )

        results = []
        for record in group.get("requests", []):
            try:
                outcome = self.run(record["id"])
            except SchemaError as exc:
                # A request that could not be sent at all -- SSL off for its
                # target, its instance gone, a stored schema that cannot be
                # honoured. It belongs in the results as a failure, not as a
                # raise that hides the rows after it.
                outcome = {"status": 0, "checked": True, "passed": False, "errors": [str(exc)]}
            outcome["id"] = record["id"]
            outcome["name"] = record.get("name")
            results.append(outcome)
        return results

    def assert_collection_passes(self, name_or_id):
        """Run a collection and raise unless **every** request met its check.

        The message names each failure with its request, because "something in
        checkout-flow broke" is not an answer anybody can act on. Unchecked
        requests are reported the same way assert_passes() treats one: they are
        not passes. Turn that off with require_check=False.
        """
        results = self.run_collection(name_or_id)

        problems = []
        for r in results:
            if not r["checked"]:
                if self.require_check:
                    problems.append("{}: nothing was checked".format(r["name"]))
                continue
            if not r["passed"]:
                problems.extend("{}: {}".format(r["name"], line) for line in r["errors"])

        if problems:
            raise ExpectationFailed('Collection "{}"'.format(name_or_id), problems)
        return results


def check(request_id, expect=_UNSET, host=None, port=None, require_check=True):
    """Module-level shortcut mirroring capture() / mock_set() / blocked():
    send one saved request and raise unless the response met its check."""
    client = SchemaClient(host=host, port=port, require_check=require_check)
    return client.assert_passes(request_id, expect)
